File: leetcode3/CheckDistancesBetweenSameLetters/a.ts

// 68ms 100.00% 45MB 100.00%
// brute force
// O(N) O(1)
function checkDistances(s: string, distances: number[]): boolean {
    let n = s.length
    let prevIdxs = new Array<number>().fill(-1);
    
    for (let i = 0; i < n; ++i) {
        let c = s.charAt(i);
        let idx = c.charCodeAt(0) - 97;
        if (prevIdxs[idx] >= 0) {
            let dist = i - prevIdxs[idx] - 1;
            if (dist != distances[idx]) {
                return false;
            }
        }
        prevIdxs[idx] = i;
    }
    return true;
};

File: leetcode3/CheckDistancesBetweenSameLetters/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2025 by iamslash

from typing import List

# 3ms 27.19% 17.13MB 100.00%
# O(N) O(1)
class Solution:
    def checkDistances(self, s: str, distance: List[int]) -> bool:
        n = len(s)
        positions = [-1] * 26

        for pos in range(n):
            idx = ord(s[pos]) - ord('a')

            if positions[idx] >= 0:
                dist = pos - positions[idx] - 1
                if distance[idx] != dist:
                    return False

            positions[idx] = pos

        return True

# 4ms 13.16% 17.23MB 99.12%
# hash map
# O(N) O(1)
class Solution:
    def checkDistances(self, s: str, distance: List[int]) -> bool:
        posMap = {}

        for i in range(len(s)):
            idx = ord(s[i]) - ord('a')
            if idx not in posMap:
                posMap[idx] = i
            else:
                if i - posMap[idx] - 1 != distances[idx]:
                    return False
        return True

# 0ms 100.00% 17.14MB 100.00%
# hash map
# O(N) O(1)
class Solution:
    def checkDistances(self, s: str, distances: List[int]) -> bool:
        positions = [-1] * 26

        for pos, char in enumerate(s):
            idx = ord(char) - ord('a')

            if positions[idx] != -1:
                if pos - positions[idx] - 1 != distances[idx]:
                    return False
            else:
                positions[idx] = pos;
                
        return True
        
def main():
    pass

if __name__ == "__main__":
    main()
